Add `head_only?` to Formula, replace `head?` in livecheck

nandahkrishna opened the pull request:
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

The `brew livecheck --installed` command is supposed to run `livecheck` for all formulae that have been installed. As `livecheck` reports new versions of software from upstream (compared to the homebrew-core version and not the system-installed version), one would expect this command to work just like `brew livecheck` does but only limits the Formulae checked.

However,
```
➜ brew livecheck arduino-cli     
arduino-cli : 0.12.1 ==> 0.12.1
➜ brew livecheck --installed                   
adns : 1.6.0 ==> 1.6.0
anime-downloader : 4.6.4 ==> 4.6.4
aom : 2.0.0 ==> 2.0.0
==> Cloning https://github.com/arduino/arduino-cli.git
Updating /Users/nanda/Library/Caches/Homebrew/arduino-cli--git
From https://github.com/arduino/arduino-cli
   f1877ef..693a045  master     -> origin/master
==> Checking out branch master
Already on 'master'
Your branch is behind 'origin/master' by 1 commit, and can be fast-forwarded.
  (use "git pull" to update your local branch)
HEAD is now at 693a045 [skip changelog] Completely document resources not inherited from referenced platform (#940)
arduino-cli : b8cf32d ==> 693a045
```
This happens because I have a `HEAD`-version of `arduino-cli` installed, despite it not being a HEAD-only formula.

This PR fixes the `--installed` flag to check only installed formulae, but report their stable version instead (if it exists).

```
➜ brew livecheck --installed
adns : 1.6.0 ==> 1.6.0
anime-downloader : 4.6.4 ==> 4.6.4
aom : 2.0.0 ==> 2.0.0
arduino-cli : 0.12.1 ==> 0.12.1
```

For a HEAD-only formula, we continue to report HEAD version updates.

One problem still persists, for HEAD-only formulae, the output of `fetch_latest_commit` is not silenced despite the `formula.head.downloader.shutup!`. While I tried looking through `Context` and `AbstractDownloadStrategy`, I wasn't able to come up with a fix. It seems like setting `@quiet = true` isn't working. Any idea why?

dawidd6 commented:
You can try adding `@quiet = false` below this line:
https://github.com/Homebrew/brew/blob/cfa03c8cce62c1b1b78ade0b20d69611cd6a89b7/Library/Homebrew/download_strategy.rb#L40

nandahkrishna commented on Library/Homebrew/dev-cmd/livecheck.rb:
```diff
@@ -51,7 +51,7 @@ def livecheck
       Tap.fetch(args.tap).formula_names.map { |name| Formula[name] }
     elsif args.installed?
-      Formula.installed
+      Formula.installed.map { |formula| Formula[formula.full_name] }
```
I think `Formula.installed` is correct in behaving the way it does. It is supposed to return what is installed – and in the example I gave, I had a `head` version installed, which is what it returns.

It's just that `livecheck` isn't a 'local' update checker, but a developer command probably used by contributors who want to bump formula versions, for example in `homebrew/core`.

The scenario this PR fixes is an edge case, where the person has the `head`-version installed (maybe to get rid of a bug fix in the latest stable version), and wants to check for the next stable version to be released with `livecheck`.

MikeMcQuaid replied:
Ok. I think at least this warrants a comment but it's not clear to the casual reader why this wouldn't return the same as `Formula.installed`. An optional argument for it may make more sense in making this more explicit?

nandahkrishna replied:
I can add a comment explaining the situation, yes. As for the argument, I'm not too sure what it would be, but I don't think there's a strong case for one. Livecheck identifies the latest upstream versions of formulae, so `--installed` just means "fetch the latest versions of formulae that I have on my system."

I don't feel too strongly about not having the option though, so I will add it if it makes more sense.

samford commented on Library/Homebrew/dev-cmd/livecheck.rb:
```diff
@@ -51,7 +51,7 @@ def livecheck
       Tap.fetch(args.tap).formula_names.map { |name| Formula[name] }
     elsif args.installed?
-      Formula.installed
+      Formula.installed.map { |formula| Formula[formula.full_name] }
```
At least to me, it seems like the main issue is that our assumptions around `#head?`aren't correct all the time. `#head?`, `#stable?`, etc. are only telling us what the `active_spec` is, so `head?` isn't a reliable way of identifying _only_ head-only formulae.

One way of resolving this (instead of modifying `Formula#installed`) is to add a `#head_only?` method, like we discussed previously. Something like:

```ruby
def head_only?
  head && !stable
end
```

When I added this to `Formula` and replaced instances of `formula.head?` in livecheck with `formula.head_only?`, this situation is resolved. That is to say:

* A formula with `stable` and `head` that's installed as `head` will do a normal check instead of the head-only check.
* An installed head-only formula will do the head-only check, as expected.

nandahkrishna replied:
Thanks Sam, this was indeed the case. I'll update the PR to include the `head_only?` method.

nandahkrishna commented:
As suggested by Sam (and discussed previously), I've added `head_only?` to the `Formula` class, which fixes this issue. Livecheck will now carry out the normal check for non-HEAD-only formulae, which was the original behaviour before we began using `formula.head?`.

I've also added a comment stating this (worded almost exactly like Sam's comment in their review) to `dev-cmd/livecheck.rb`. Do let me know if anything has to be modified. Thanks!

samford commented on Library/Homebrew/dev-cmd/livecheck.rb:
```diff
@@ -51,6 +51,8 @@ def livecheck
     formulae_to_check = if args.tap
       Tap.fetch(args.tap).formula_names.map { |name| Formula[name] }
     elsif args.installed?
+      # Livecheck will carry out a normal check for formulae having `stable` and `head` that are
+      # installed as `head`, rather than a HEAD-only check.
```
It may just be me but this doesn't feel like the best place for this comment. It would be nice to explain the idea behind this (i.e., livecheck will only use the installed `head` version for head-only formulae) but I don't think it makes sense here.

It may be better to explain this where the `current` variable is defined, as that's where the behavior comes into play (I'll add a comment for this below). I think that using the `stable` version for comparison except with head-only formulae is predictable and in-line with livecheck's normal behavior, so I don't think it's necessary to explain this in relation to the `--installed` switch.

samford commented on Library/Homebrew/livecheck/livecheck.rb:
```diff
@@ -78,19 +78,19 @@ def livecheck_formulae(formulae_to_check, args)
+        formula.head.downloader.shutup! if formula.head_only?
 
-        current = if formula.head?
+        current = if formula.head_only?
```
```suggestion
        # Use the `stable` version for comparison except for installed
        # head-only formulae. A formula with `stable` and `head` that's
        # installed using `--head` will still use the `stable` version for
        # comparison.
        current = if formula.head_only?
```

If we need to explain this behavior, this may make more sense than the above. However, I think the code speaks for itself and even this comment may not be necessary.

I would wait to integrate this change until others weigh in on whether they think we need comments to explain this idea (here or elsewhere).

nandahkrishna replied:
@MikeMcQuaid any thoughts on this?

MikeMcQuaid replied:
@samford's explanation seems good 👍🏻

samford reviewed (commented):
> changed the value of `current` to `formula.stable.version`, and that seems to solve the issue

I tested this and can confirm that the reported version for a formula with `stable` and `head` that's installed using `--head` is the HEAD commit hash when running `brew livecheck --installed`. Using `formula.stable.version` instead of `formula.version` addresses this and makes sense to me 👍

nandahkrishna commented:
Sorry for the delay, was slightly occupied with some other work. With the latest push:
* The comment was reworded and relocated (based on Sam's review).
* It seems `shutup!` is only used by `livecheck`, or that's what it seems like to me from a `grep -R`. Also, `quiet?` only checks if the current Context has `@quiet` set to true, which seems to happen only when the `--quiet` flag is passed. Taking inspiration from `mktemp.rb` and Dawid's suggestion, I've defined a new `quiet?` specific to `AbstractDownloadStrategy`. Now, there is no output from `fetch_last_commit` during `livecheck` runs.
